cataclysmbnteam/Cataclysm-BN
# Strength affects stamina cost of melee, rescale dex effect on movecost

#### Summary



SUMMARY: Balance "Above-average strength steadily reduces stamina cost of melee attacks, rework how dexterity affects movecost of melee attacks to match"

#### Purpose of change



I was looking into melee code yet again and saw two things: first, melee time cost gets a hilariously useless bonus from dexterity, reducing melee movecost by basically a centisecond for every 2 points of dexterity. This has the fun combo of being useless at any sane amount of dexterity, rewarding faster weapons more than slower weapons (bot be much since amount is pathetically low but still), and of course enabling timefreeze shenanigans with absurdly-high amounts of debugged-in dexterity.

Meanwhile stamina cost I noticed has a thing indicating that it USED to divide the effective weight of an item by strength to determine stamina cost, and instead uses a fixed ratio.

#### Describe the solution



1. In melee.cpp, per Coolthulhu's feedback I set `Character::attack_cost` to stay an additive bonus instead of using the multiplicative method shown below, simply making it scale at a rate that makes dexterity actually matter a bit more.
2. Meanwhile in `Character::melee_attack`, stamina cost now takes strength into account. Per feedback it will penalize you if you have below-average strength, with it currently maxing out at a 1.25x multiplier at worst. It's tuned to currently produce a 1x multiplier (thus, no change) at the current average of 8 strength, and infinitely trends closer to a maximum of 0.75x as your strength goes beyond that. Iit creates this multiplier by adding 0.75 to `strbonus`, which is why it can trend towards 0.75x as the maximum possible multiple. It's tuned so that feeding it 8 strength will output `strbonus` of 0.25, thus added to that 0.75 it becomes the expected 1x multiplier, while zero strength will output a `strbonus` of 0.5 (thus makin the multiplier 1.25x).
3. In addition, Kheir pointed out it doesn't need the "multiply it by minus one" thing to output a negative result, and showed a better way to do that.

For comparison, the strength scaling the code comment alludes to implies that having 20 strength back in the ancient days make the stamina cost 40% of what it'd be at 8 strength, while here the result will be ~89% instead.

#### Describe alternatives you've considered



1. It's also been suggested that we target 10 as "average" strength instead of 8. I could but I feel like it'd be saner to target the character baseline as where things would return a 1x multiplier, as that's seen as the normal default.
2. Making the range 1.333 down to 0.666 instead. Could work, but I feel like we should go easy with both how good it can get at high strength and how punishing it can get with low strength, until we know for sure how it plays. Remember, pain can tank your speed on top of lowering stats down to zero. It would also make the already kinda fucking cursed math required to determine `strbonus` without any if/else hackery or `std::min` look even worse, as we'd need to replace the `0.75` with `( 2 / 3 )`for it to reliably return 1x at whatever baseline strength value we pick. Just using `0.666f` or `0.6667` might not cut it.
3. Making dexterity use multiplicative logic like stamina cost does. Coolthulhu advised against it on the basis that dex should benefit small light weapons more than big heavy ones, which an additive bonus will do where a multiplicative one would favor heavy ones instead. Personally I feel it'd be fine to do either, if anything small light weapons are already kinda the go-to players always insist on using, but it works either way.
4. Waiting on the stamina cost bit at least and tying that in with athletics skill if we ever do that. I feel like we can probably do it like this for now and rework whenever the hassle of athletics skill is out of the way.
5. Checking for any other big obvious time and stamina penalties that lack stat scaling, and/or have very pathetic scaling like attack movecost did.

#### Testing



First, compiled in upload branch and tested the values of a zweihander with a naked, straight-8s character at full stamina and no skills. Move cost was 164 and stamina cost according to debug mode was 248. Next:
1. Compiled and load-tested.
2. Repeated the above test, move cost was now 160 (4 more since 8 dex, as expected), stamina cost was still 248.
3. Tested with zero strength, stamina cost was 310 (expected value 310).
4. Tested at 20 strength, stamina cost was 221 (expected value ~221.43).
6. Tested at 1000 strength, stamina cost was 186 (expected value ~186.98).
7. Next tested movecost at zero dexterity, movecost was 168 (close but not quite listed item movecost, possibly a tiny error in the math unrelated to `dexbonus` stuff?).
8. Set dex to 20, movecost was 148.
9. Set dex to 1000, movecost reached the floor of 25.
10. Checked affected file for astyle.

#### Additional context
